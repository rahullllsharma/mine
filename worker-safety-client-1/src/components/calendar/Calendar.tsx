import type { CriticalActivityType } from "../../container/report/dailyInspection/types";
import { useEffect, useState } from "react";
import {
  convertDateToString,
  getDate,
  getDayRange,
  getDefaultDate,
  isDateWithinRange,
  isFollowingDate,
  isPreviousDate,
} from "@/utils/date/helper";
import { RiskLevel } from "../riskBadge/RiskLevel";
import CalendarDayButton from "./calendarButton/calendarDayButton/CalendarDayButton";
import CalendarNavButton from "./calendarButton/calendarNavButton/CalendarNavButton";
import DateSelector from "./dateSelector/DateSelector";

export type CalendarProps = {
  startDate: string;
  endDate: string;
  onDateSelect: (date: string) => void;
  defaultDate?: string;
  isCritical?: boolean;
  dateRangeNavigator?: (initialDate: string, finalDate: string) => void;
  criticalActivityData?: CriticalActivityType[];
  riskLevel?: RiskLevel;
};

export default function Calendar({
  startDate,
  endDate,
  onDateSelect,
  defaultDate = "",
  dateRangeNavigator,
  criticalActivityData,
  riskLevel = RiskLevel.UNKNOWN,
}: CalendarProps): JSX.Element {
  const today = convertDateToString();
  const [selectedDate, setSelectedDate] = useState(
    getDefaultDate(startDate, endDate, defaultDate || today)
  );
  const [responsiveNavHelper, setResponsiveNavHelper] = useState(3); //middle item
  const [dateRange, setDateRange] = useState(
    getDayRange(getDefaultDate(startDate, endDate, today), 3, 3)
  );

  const isDayDisabled = (date: string) =>
    !isDateWithinRange(startDate, endDate, date);

  const isTodayDisabled = !isDateWithinRange(startDate, endDate, today);
  const isLastWeekDisabled = isPreviousDate(dateRange[0], startDate);
  const isPreviousDayDisabled = isPreviousDate(selectedDate, startDate);
  const isNextWeekDisabled = isFollowingDate(dateRange[6], endDate);
  const isNextDayDisabled = isFollowingDate(selectedDate, endDate);

  const setDate = (date: string): void => {
    setSelectedDate(date);
    onDateSelect(date);
  };

  const todayClickHandler = (): void => {
    if (!isDateWithinRange(dateRange[0], dateRange[6], today)) {
      setDateRange(getDayRange(today, 3, 3));
    }
    setDate(today);
    setResponsiveNavHelper(3);
  };

  const navigateToPreviousDay = (): void => {
    if (responsiveNavHelper === 0) {
      setDateRange(getDayRange(getDate(selectedDate, -1), 6, 0));
      setResponsiveNavHelper(6);
    } else {
      setResponsiveNavHelper(prevState => prevState - 1);
    }
    setDate(getDate(selectedDate, -1));
  };

  const navigateToNextDay = (): void => {
    if (responsiveNavHelper === 6) {
      setDateRange(getDayRange(getDate(selectedDate, 1), 0, 6));
      setResponsiveNavHelper(0);
    } else {
      setResponsiveNavHelper(prevState => prevState + 1);
    }
    setDate(getDate(selectedDate, 1));
  };

  const navigateToLastWeek = (): void => {
    const date = getDate(dateRange[0], -1);
    setDateRange(getDayRange(date, 6, 0));
    setDate(date);
    setResponsiveNavHelper(6);
  };

  const navigateToNextWeek = (): void => {
    const date = getDate(dateRange[6], 1);
    setDateRange(getDayRange(date, 0, 6));
    setSelectedDate(date);
    setDate(date);
    setResponsiveNavHelper(0);
  };

  useEffect(() => {
    dateRangeNavigator ? dateRangeNavigator(dateRange[0], dateRange[6]) : null;
  }, [dateRange]);

  return (
    <>
      <DateSelector
        className="mb-3"
        date={selectedDate}
        isTodayDisabled={isTodayDisabled}
        isPreviousDateDisabled={isPreviousDayDisabled}
        isNextDateDisabled={isNextDayDisabled}
        onTodayClicked={todayClickHandler}
        onPreviousDateClicked={navigateToPreviousDay}
        onNextDateClicked={navigateToNextDay}
      />
      <div className="md:flex justify-between hidden">
        <CalendarNavButton
          className="lg:mr-12 mr-4"
          icon="chevron_big_left"
          isDisabled={isLastWeekDisabled}
          onClick={navigateToLastWeek}
        />
        <div className="flex flex-1 lg:justify-center justify-between">
          {dateRange.map(date => (
            <CalendarDayButton
              className="lg:mr-8 last:mr-0"
              key={date}
              date={date}
              isDisabled={isDayDisabled(date)}
              isActive={selectedDate === date}
              onClick={() => setDate(date)}
              isCritical={
                (
                  criticalActivityData?.find(
                    (data: CriticalActivityType) => data.date === date
                  ) || {}
                ).isCritical || false
              }
              riskLevel={riskLevel}
            />
          ))}
        </div>
        <CalendarNavButton
          className="lg:ml-12 ml-4"
          icon="chevron_big_right"
          isDisabled={isNextWeekDisabled}
          onClick={navigateToNextWeek}
        />
      </div>
    </>
  );
}
