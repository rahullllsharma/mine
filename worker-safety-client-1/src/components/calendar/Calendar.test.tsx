import { fireEvent, render, screen } from "@testing-library/react";
import {
  convertDateToString,
  getDate,
  getDayAndWeekdayFromDate,
  getFormattedDate,
} from "@/utils/date/helper";
import Calendar from "./Calendar";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: true,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const mockOnDateSelect = jest.fn();

jest.useFakeTimers();
jest.setSystemTime(new Date("2022-01-24").valueOf());

describe(Calendar.name, () => {
  describe("when it renders", () => {
    const startDate = "2022-01-01";
    const endDate = "2030-01-20";

    it("should have 7 calendar day buttons and two navigation buttons", () => {
      const { asFragment } = render(
        <Calendar
          startDate={startDate}
          endDate={endDate}
          onDateSelect={mockOnDateSelect}
          defaultDate={startDate}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should have the current day as a default when the current day is within the project range", () => {
      const today = convertDateToString(new Date());

      render(
        <Calendar
          startDate={startDate}
          endDate={endDate}
          onDateSelect={mockOnDateSelect}
          defaultDate={today}
        />
      );

      const [day, weekday] = getDayAndWeekdayFromDate(today);
      const buttonElement = screen.getByRole("button", {
        name: `${day} ${weekday}`,
      });
      expect(buttonElement).toHaveAttribute("data-is-today");
    });

    it("should have the endDate as default when the current day is after the project end date", () => {
      const date = convertDateToString("2021-01-25");

      render(
        <Calendar
          startDate="2021-01-15"
          endDate="2021-01-25"
          onDateSelect={mockOnDateSelect}
          defaultDate={date}
        />
      );

      const [day, weekday] = getDayAndWeekdayFromDate(date);
      const buttonElement = screen.getByRole("button", {
        name: `${day} ${weekday}`,
      });
      expect(buttonElement).toHaveAttribute("data-is-selected");
    });

    it("should have the startDate as default when the current day is before the project start date", () => {
      const date = convertDateToString("2030-01-15");

      render(
        <Calendar
          startDate="2030-01-15"
          endDate="2030-01-25"
          onDateSelect={mockOnDateSelect}
          defaultDate={date}
        />
      );

      const [day, weekday] = getDayAndWeekdayFromDate(date);
      const buttonElement = screen.getByRole("button", {
        name: `${day} ${weekday}`,
      });
      expect(buttonElement).toHaveAttribute("data-is-selected");
    });
  });

  describe("when dealing with the today button", () => {
    it("should select the current day when it's clicked", () => {
      const today = convertDateToString(new Date());

      render(
        <Calendar
          startDate="2022-01-15"
          endDate="2030-01-25"
          onDateSelect={mockOnDateSelect}
          defaultDate={today}
        />
      );
      const nextDate = convertDateToString(getDate(new Date(), 1));
      const [nextDay, nextWeekday] = getDayAndWeekdayFromDate(nextDate);
      const buttonElement = screen.getByRole("button", {
        name: `${nextDay} ${nextWeekday}`,
      });
      fireEvent.click(buttonElement);
      fireEvent.click(screen.getByRole("button", { name: /today/i }));

      const [day, weekday] = getDayAndWeekdayFromDate(today);
      const todayElement = screen.getByRole("button", {
        name: `${day} ${weekday}`,
      });
      expect(todayElement).toHaveAttribute("data-is-today");
    });

    it("should select the current day when it's clicked and update the calendar range", () => {
      render(
        <Calendar
          startDate="2022-01-15"
          endDate="2030-01-25"
          onDateSelect={mockOnDateSelect}
          defaultDate="2022-01-15"
        />
      );
      fireEvent.click(
        screen.getByRole("button", {
          name: /last week/i,
        })
      );
      fireEvent.click(screen.getByRole("button", { name: /today/i }));
      const date = convertDateToString(new Date());
      const [day, weekday] = getDayAndWeekdayFromDate(date);
      const todayElement = screen.getByRole("button", {
        name: `${day} ${weekday}`,
      });
      expect(todayElement).toHaveAttribute("data-is-today");
    });

    it("should be disabled if the current day is outside the calendar dates", () => {
      render(
        <Calendar
          startDate="2021-01-15"
          endDate="2021-01-25"
          onDateSelect={mockOnDateSelect}
          defaultDate="2022-01-15"
        />
      );
      const buttonElement = screen.getByRole("button", { name: /today/i });
      expect(buttonElement).toBeDisabled();
    });
  });

  describe("when the calendar day buttons are pressed", () => {
    const startDate = "2022-01-15";
    const endDate = "2030-01-25";

    it("should switch to another active element when clicked", () => {
      const today = convertDateToString(getDate(new Date(), 1));

      render(
        <Calendar
          startDate={startDate}
          endDate={endDate}
          onDateSelect={mockOnDateSelect}
          defaultDate={today}
        />
      );

      const [day, weekday] = getDayAndWeekdayFromDate(today);
      const buttonElement = screen.getByRole("button", {
        name: `${day} ${weekday}`,
      });
      fireEvent.click(buttonElement);
      expect(buttonElement).toHaveAttribute("data-is-selected");
    });

    it("should remove the selected styling from the current day when another day is selected", () => {
      const today = convertDateToString(new Date());

      render(
        <Calendar
          startDate={startDate}
          endDate={endDate}
          onDateSelect={mockOnDateSelect}
          defaultDate={today}
        />
      );
      const date = convertDateToString(getDate(new Date(), 1));
      const [day, weekday] = getDayAndWeekdayFromDate(date);
      const buttonElement = screen.getByRole("button", {
        name: `${day} ${weekday}`,
      });
      fireEvent.click(buttonElement);

      const [todayDay, todayWeekday] = getDayAndWeekdayFromDate(today);
      const todayElement = screen.getByRole("button", {
        name: `${todayDay} ${todayWeekday}`,
      });

      expect(todayElement).not.toHaveAttribute("data-is-selected");
    });
  });

  describe("when the calendar navigation buttons are pressed", () => {
    describe('when "Last week" button is pressed', () => {
      const startDate = "2022-01-10";
      const endDate = "2022-01-20";

      beforeEach(() => {
        render(
          <Calendar
            startDate={startDate}
            endDate={endDate}
            onDateSelect={mockOnDateSelect}
            defaultDate={startDate}
          />
        );
      });

      it("should navigate to the previous week", () => {
        const buttonElement = screen.getByRole("button", {
          name: /last week/i,
        });
        fireEvent.click(buttonElement);

        const [d1, w1] = getDayAndWeekdayFromDate("2022-01-10");
        const [d2, w2] = getDayAndWeekdayFromDate("2022-01-13");

        screen.getByRole("button", { name: `${d1} ${w1}` });
        screen.getByRole("button", { name: `${d2} ${w2}` });
      });

      it("should disable the button when no more weeks are available", () => {
        const buttonElement = screen.getByRole("button", {
          name: /last week/i,
        });
        fireEvent.click(buttonElement);

        expect(buttonElement).toBeDisabled();
      });
    });

    describe('when "Next week" is pressed', () => {
      const startDate = "2030-01-15";
      const endDate = "2030-01-25";

      beforeEach(() => {
        render(
          <Calendar
            startDate={startDate}
            endDate={endDate}
            onDateSelect={mockOnDateSelect}
            defaultDate={startDate}
          />
        );
      });

      it("should navigate to the next week", () => {
        const buttonElement = screen.getByRole("button", {
          name: /next week/i,
        });
        fireEvent.click(buttonElement);

        const [d1, w1] = getDayAndWeekdayFromDate("2030-01-19");
        const [d2, w2] = getDayAndWeekdayFromDate("2030-01-25");

        screen.getByRole("button", { name: `${d1} ${w1}` });
        screen.getByRole("button", { name: `${d2} ${w2}` });
      });

      it("should disable the button when no more weeks are available", () => {
        const buttonElement = screen.getByRole("button", {
          name: /next week/i,
        });
        fireEvent.click(buttonElement);

        expect(buttonElement).toBeDisabled();
      });
    });

    describe("when on smaller screens", () => {
      const today = convertDateToString(new Date());
      const startDate = "2022-01-15";
      const endDate = "2030-01-25";

      beforeEach(() => {
        render(
          <Calendar
            startDate={startDate}
            endDate={endDate}
            onDateSelect={mockOnDateSelect}
            defaultDate={today}
          />
        );
      });
      it("should navigate to the previous day when the left arrow is pressed", () => {
        const buttonElements = screen.getAllByRole("button", {
          hidden: true,
          name: "",
        });
        fireEvent.click(buttonElements[0]);

        const date = getDate(new Date(), -1);
        screen.getByText(getFormattedDate(date, "long"));
      });

      it("should navigate to the next day when the right arrow is pressed", () => {
        const buttonElements = screen.getAllByRole("button", {
          hidden: true,
          name: "",
        });
        fireEvent.click(buttonElements[1]);

        const date = getDate(new Date(), 1);
        screen.getByText(getFormattedDate(date, "long"));
      });

      it("should update the date range when the left arrow is pressed and user reaches the previous week", () => {
        const buttonElements = screen.getAllByRole("button", { hidden: true });
        const hiddenButtons = buttonElements.filter(
          btn => btn.textContent === ""
        );

        for (let i = 0; i < 4; i++) {
          fireEvent.click(hiddenButtons[0]);
        }
        const firstDate = convertDateToString(getDate(new Date(), -10));
        const lastDate = convertDateToString(getDate(new Date(), -4));

        const [d1, w1] = getDayAndWeekdayFromDate(firstDate);
        const [d2, w2] = getDayAndWeekdayFromDate(lastDate);

        screen.getByRole("button", { name: `${d1} ${w1}` });
        screen.getByRole("button", { name: `${d2} ${w2}` });
      });

      it("should update the date range when the left arrow is pressed and user reaches the next week", () => {
        const buttonElements = screen.getAllByRole("button", { hidden: true });
        const hiddenButtons = buttonElements.filter(
          btn => btn.textContent === ""
        );

        for (let i = 0; i < 4; i++) {
          fireEvent.click(hiddenButtons[1]);
        }
        const firstDate = convertDateToString(getDate(new Date(), 4));
        const lastDate = convertDateToString(getDate(new Date(), 10));

        const [d1, w1] = getDayAndWeekdayFromDate(firstDate);
        const [d2, w2] = getDayAndWeekdayFromDate(lastDate);

        screen.getByRole("button", { name: `${d1} ${w1}` });
        screen.getByRole("button", { name: `${d2} ${w2}` });
      });
    });
  });
});
