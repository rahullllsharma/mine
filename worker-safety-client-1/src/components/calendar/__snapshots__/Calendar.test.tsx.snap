// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar when it renders should have 7 calendar day buttons and two navigation buttons 1`] = `
<DocumentFragment>
  <div
    class="flex items-center mb-3"
  >
    <button
      class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed text-sm rounded-md font-semibold bg-brand-urbint-10 border border-brand-urbint-40 shadow-5 hover:shadow-10 text-neutral-shade-75"
      type="button"
    >
      <div
        class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 px-2 py-1"
      >
        <span
          class="mx-1 truncate"
        >
          Today
        </span>
      </div>
    </button>
    <p
      class="text-lg md:text-xl ml-4 text-neutral-shade-100 truncate"
    >
      January 1, 2022
    </p>
    <div
      class="flex ml-4 md:hidden text-neutral-shade-75"
    >
      <button
        class="text-xl text-neutral-shade-75 disabled:text-neutral-shade-38 disabled:cursor-not-allowed"
        disabled=""
      >
        <i
          aria-hidden="true"
          class="ci-chevron_big_left"
        />
      </button>
      <button
        class="text-xl text-neutral-shade-75 disabled:text-neutral-shade-38 disabled:cursor-not-allowed ml-4"
      >
        <i
          aria-hidden="true"
          class="ci-chevron_big_right"
        />
      </button>
    </div>
  </div>
  <div
    class="md:flex justify-between hidden"
  >
    <button
      class="h-20 w-20 flex flex-col items-center justify-center lg:mr-12 mr-4 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
    >
      <i
        aria-hidden="true"
        class="ci-chevron_big_left text-xl leading-5 text-neutral-shade-75"
      />
      <span
        class="mt-1 text-xs font-base text-neutral-shade-100"
      >
        Last week
      </span>
    </button>
    <div
      class="flex flex-1 lg:justify-center justify-between"
    >
      <button
        class="h-20 w-20 flex flex-col items-center justify-center lg:mr-8 last:mr-0 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
      >
        <span
          class="text-base font-semibold text-brand-urbint-60"
        >
          21
        </span>
        <div
          class="flex"
        >
          <span
            class="mt-1 text-tiny font-medium text-neutral-shade-100"
          >
            Fri
          </span>
        </div>
      </button>
      <button
        class="h-20 w-20 flex flex-col items-center justify-center lg:mr-8 last:mr-0 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
      >
        <span
          class="text-base font-semibold text-brand-urbint-60"
        >
          22
        </span>
        <div
          class="flex"
        >
          <span
            class="mt-1 text-tiny font-medium text-neutral-shade-100"
          >
            Sat
          </span>
        </div>
      </button>
      <button
        class="h-20 w-20 flex flex-col items-center justify-center lg:mr-8 last:mr-0 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
      >
        <span
          class="text-base font-semibold text-brand-urbint-60"
        >
          23
        </span>
        <div
          class="flex"
        >
          <span
            class="mt-1 text-tiny font-medium text-neutral-shade-100"
          >
            Sun
          </span>
        </div>
      </button>
      <button
        class="h-20 w-20 flex flex-col items-center justify-center lg:mr-8 last:mr-0 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-urbint-40"
        data-is-today="true"
      >
        <span
          class="text-base font-semibold text-brand-urbint-40"
        >
          24
        </span>
        <div
          class="flex"
        >
          <span
            class="mt-1 text-tiny font-medium text-neutral-shade-100"
          >
            Mon
          </span>
        </div>
      </button>
      <button
        class="h-20 w-20 flex flex-col items-center justify-center lg:mr-8 last:mr-0 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
      >
        <span
          class="text-base font-semibold text-brand-urbint-60"
        >
          25
        </span>
        <div
          class="flex"
        >
          <span
            class="mt-1 text-tiny font-medium text-neutral-shade-100"
          >
            Tue
          </span>
        </div>
      </button>
      <button
        class="h-20 w-20 flex flex-col items-center justify-center lg:mr-8 last:mr-0 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
      >
        <span
          class="text-base font-semibold text-brand-urbint-60"
        >
          26
        </span>
        <div
          class="flex"
        >
          <span
            class="mt-1 text-tiny font-medium text-neutral-shade-100"
          >
            Wed
          </span>
        </div>
      </button>
      <button
        class="h-20 w-20 flex flex-col items-center justify-center lg:mr-8 last:mr-0 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
      >
        <span
          class="text-base font-semibold text-brand-urbint-60"
        >
          27
        </span>
        <div
          class="flex"
        >
          <span
            class="mt-1 text-tiny font-medium text-neutral-shade-100"
          >
            Thu
          </span>
        </div>
      </button>
    </div>
    <button
      class="h-20 w-20 flex flex-col items-center justify-center lg:ml-12 ml-4 disabled:opacity-30 disabled:cursor-not-allowed border rounded-xl border rounded-xl border-brand-gray-40"
    >
      <i
        aria-hidden="true"
        class="ci-chevron_big_right text-xl leading-5 text-neutral-shade-75"
      />
      <span
        class="mt-1 text-xs font-base text-neutral-shade-100"
      >
        Next week
      </span>
    </button>
  </div>
</DocumentFragment>
`;
